url90b6689baa_adsonrocha_PD321_tgz-pJ8-MyFuzzyLiteClassJ8/src/com/fuzzylite/rule/Rule.java

```diff
@@ -0,0 +1,173 @@
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+package com.fuzzylite.rule;
+
+import com.fuzzylite.Engine;
+import com.fuzzylite.Op;
+import static com.fuzzylite.Op.str;
+import com.fuzzylite.norm.SNorm;
+import com.fuzzylite.norm.TNorm;
+import java.util.StringTokenizer;
+
+/**
+ *
+ * @author jcrada
+ */
+public class Rule {
+
+    public static final String FL_IF = "if";
+    public static final String FL_IS = "is";
+    public static final String FL_EQUALS = "=";
+    public static final String FL_THEN = "then";
+    public static final String FL_AND = "and";
+    public static final String FL_OR = "or";
+    public static final String FL_WITH = "with";
+    protected Antecedent antecedent;
+    protected Consequent consequent;
+    protected double weight = 1.0;
+    protected String text;
+
+    public Rule() {
+    }
+
+    public double activationDegree(TNorm conjunction, SNorm disjunction) {
+        return weight * this.antecedent.activationDegree(conjunction, disjunction);
+    }
+
+    public void activate(double activationDegree, TNorm activation) {
+        this.consequent.modify(activationDegree, activation);
+    }
+
+    public static Rule parse(String rule, Engine engine) {
+        Rule result = new Rule();
+        result.setText(rule);
+        StringTokenizer tokenizer = new StringTokenizer(rule);
+        String token;
+        String strAntecedent = "";
+        String strConsequent = "";
+
+        final byte S_NONE = 0, S_IF = 1, S_THEN = 2, S_WITH = 3, S_END = 4;
+        byte state = S_NONE;
+
+        while (tokenizer.hasMoreTokens()) {
+            token = tokenizer.nextToken();
+
+            switch (state) {
+                case S_NONE:
+                    if (Rule.FL_IF.equals(token)) {
+                        state = S_IF;
+                    } else {
+                        throw new RuntimeException(String.format(
+                                "[syntax error] expected keyword <%s>, but found <%d> "
+                                + "in rule: %s", Rule.FL_IF, token, rule));
+                    }
+                    break;
+
+                case S_IF:
+                    if (Rule.FL_THEN.equals(token)) {
+                        state = S_THEN;
+                    } else {
+                        strAntecedent += token + " ";
+                    }
+                    break;
+                case S_THEN:
+                    if (Rule.FL_WITH.equals(token)) {
+                        state = S_WITH;
+                    } else {
+                        strConsequent += token + " ";
+                    }
+                    break;
+                case S_WITH:
+                    try {
+                        result.setWeight(Op.toDouble(token));
+                        state = S_END;
+                    } catch (NumberFormatException ex) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)